src/ray/core_worker/task_state_buffer.cc

```diff
@@ -0,0 +1,175 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/task_state_buffer.h"
+
+namespace ray {
+namespace core {
+
+namespace worker {
+
+TaskStateBuffer::TaskStateBuffer(instrumented_io_context &io_service,
+                                 const std::shared_ptr<gcs::GcsClient> &gcs_client)
+    : io_service_(io_service), periodical_runner_(io_service_), gcs_client_(gcs_client) {
+  periodical_runner_.RunFnPeriodically([this] { FlushEvents(); },
```

Comment:
Btw ,if you use `RunFnPeriodically `, the method can be called after this class is destructed. So you should make sure to handle that case within `FlushEvents`

Reply:
hmm maybe instead of periodically flushing it, we can flush upon reply? (with some delay). That could be a better backpressure policy? Doesn't need to be done from the prototype.

Reply:
Ah that's good to know! Yeah, making it synchronous with reply could be a solution.